lib/ModelMixins/TableMixin.ts

```diff
@@ -53,19 +53,63 @@ function TableMixin<T extends Constructor<Model<TableTraits>>>(Base: T) {
     get hasTableMixin() {
       return true;
     }
-    /**
-     * The raw data table in column-major format, i.e. the outer array is an
-     * array of columns.
-     */
+
+    // Always use the getter and setter for this
     @observable
-    dataColumnMajor: string[][] | undefined;
+    protected _dataColumnMajor: string[][] | undefined;
 
     /**
      * The list of region providers to be used with this table.
      */
     @observable
     regionProviderList: RegionProviderList | undefined;
 
+    /**
+     * The raw data table in column-major format, i.e. the outer array is an
+     * array of columns.
+     */
+    @computed
+    get dataColumnMajor(): string[][] | undefined {
+      const dataColumnMajor =
+        this._dataColumnMajor === undefined
+          ? this._dataColumnMajor
+          : toJS(this._dataColumnMajor); // clone it so we can modify it without breaking observable rules
+      if (
+        this.removeDuplicateRows &&
+        dataColumnMajor !== undefined &&
+        dataColumnMajor.length >= 1
+      ) {
+        // De-duplication is slow and memory expensive, so should be avoided if possible.
+        const duplicateString = "TERRIAJS:DUPLICATE";
+        const seenRows = new Set();
+        let hasDuplicates = false;
+        for (let i = 0; i < dataColumnMajor[0].length; i++) {
+          const row = dataColumnMajor.map(col => col[i]).join();
+          if (seenRows.has(row)) {
+            // Mark all the cells in this row for deletion
+            dataColumnMajor.forEach(col => (col[i] = duplicateString));
+            hasDuplicates = true;
+          } else {
+            seenRows.add(row);
+          }
+        }
+        if (hasDuplicates) {
+          dataColumnMajor.forEach(
+            // Remove all the cells marked as duplicates
+            (col, idx) =>
+              (dataColumnMajor[idx] = col.filter(
+                cell => cell !== duplicateString
+              ))
+          );
+        }
+      }
+      return dataColumnMajor;
```

Comment:
I don't think you should have to clone `this._dataColumnMajor`, or change values in `dataColumnMajor`.

Maybe a better approach would be to have an array of row indices which need to be removed, and then return:

```ts
return this._dataColumnMajor.filter((row, idx) => rowIndicesToRemove.includes(idx))
```

Reply:
Also, I think it is important that if `removeDuplicateRows` is `false`, then return `this._dataColumnMajor` immediately.

Your current approach clones `this._dataColumnMajor` regardless of  what `removeDuplicateRows` is set to